src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -124,17 +124,30 @@ public partial class ProjectDependencyGraph
             return builder?.ToImmutable() ?? map;
         }
 
-        internal ProjectDependencyGraph WithProjectReferences(ProjectId projectId, IEnumerable<ProjectId> referencedProjectIds)
+        internal ProjectDependencyGraph WithProjectReferences(ProjectId projectId, IReadOnlyList<ProjectReference> projectReferences)
         {
             Contract.ThrowIfFalse(_projectIds.Contains(projectId));
 
-            // This method we can't optimize very well: changing project references arbitrarily could invalidate pretty much anything. The only thing we can reuse is our
-            // actual map of project references for all the other projects, so we'll do that
-            var referencedProjects = referencedProjectIds.ToImmutableHashSet();
-            var referencesMap = referencedProjects.IsEmpty
-                ? _referencesMap.Remove(projectId)
-                : _referencesMap.SetItem(projectId, referencedProjects);
-            return new ProjectDependencyGraph(_projectIds, referencesMap);
+            // This method we can't optimize very well: changing project references arbitrarily could invalidate pretty much anything.
+            // The only thing we can reuse is our actual map of project references for all the other projects, so we'll do that.
+
+            ImmutableHashSet<ProjectId> projectIds;
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? referencesMap;
+            if (projectReferences.IsEmpty())
+            {
+                referencesMap = _referencesMap.Remove(projectId);
+                projectIds = _projectIds;
+            }
+            else
+            {
+                var referencedProjectIds = projectReferences.Select(p => p.ProjectId).ToImmutableHashSet();
+                referencesMap = _referencesMap.SetItem(projectId, referencedProjectIds);
+
+                // add referenced project ids to handle referenced to projects outside of the solution:
+                projectIds = _projectIds.Union(referencedProjectIds);
```

Comment:
I believe the intent was the main IDs collection doesn't contain the ones that were floating outside, and those were supposed to be filtered. @sharwell am I mis-remembering?

Reply:
Without this the invariant validation fails, so either the validation is wrong or the _projectIds are wrong.

Reply:
This invariant was added in 53b0c2a7d282e6113f2694915fcc950406998d1b and validated here:

https://github.com/dotnet/roslyn/blob/9e786147b8cb884af454db081bb747a5bd36a086/src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs#L458-L461

Reply:
I'd be OK with throwing here if the solution didn't already contain the target of the reference.

Reply:
> That can only occur via misuse of a public API.

The public API explicitly caters to this, namely the distinction between "Project.ProjectReferences" vs. "Project.AllProjectReferences". I don't now if this is a useful feature anymore (at one point MSBuildWorkspace used it I think) but I'd say if we're going to change that let's change that specific thing in our 16.7 feature branch just to get that behavior change in early in a cycle so we can get time to find out what we broke.

> Without this the invariant validation fails, so either the validation is wrong or the _projectIds are wrong.

I dug a bit further: so the assumption in the existing implementation (at least at one point) was that the dependency graph isn't told about project references to non-existent projects. Then during project addition we attempt to fix those up:

https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs#L429-L441

So WithProjectReferences here should be filtering out the references that go into the ProjectDependencyGraph, since it's fixed up in that AddProject code I linked to above. Or alternately we change how it should work and delete the code in AddProject.

Reply:
So, what is the right fix for 16.6? Should I undo this and skip tests that add the references to projects that are not in the solution?